driver/main.c

```diff
@@ -70,6 +140,261 @@ PetscErrorCode GetCurrentData(PetscScalar *data_ptr, PetscInt ndata, PetscReal c
   PetscFunctionReturn(PETSC_SUCCESS);
 }
 
+// loads the binary data in a Vec
+static PetscErrorCode OpenSpatiallyHomogenousRainData(HomogeneousRainData *homogeneous_rain) {
+  PetscFunctionBegin;
+
+  PetscCall(OpenData(homogeneous_rain->filename, &homogeneous_rain->data_vec, &homogeneous_rain->ndata));
+  PetscCall(VecGetArray(homogeneous_rain->data_vec, &homogeneous_rain->data_ptr));
+  homogeneous_rain->cur_idx                = -1;
+  homogeneous_rain->prev_idx               = -1;
+  homogeneous_rain->temporally_interpolate = PETSC_FALSE;
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+// close and destroys the Vec
+static PetscErrorCode CloseSpatiallyHomogeneousRainData(HomogeneousRainData *homogeneous_rain) {
+  PetscFunctionBegin;
+
+  PetscCall(VecRestoreArray(homogeneous_rain->data_vec, &homogeneous_rain->data_ptr));
+  PetscCall(VecDestroy(&homogeneous_rain->data_vec));
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+// set a constant rainfall for all grid cells
+PetscErrorCode SetConstantRainfall(PetscInt ncells, PetscReal rain[ncells]) {
+  PetscFunctionBegin;
+
+  // apply a 1 mm/hr rain over the entire domain
+  PetscReal rain_rate             = 1.0;
+  PetscReal mm_per_hr_2_m_per_sec = 1.0 / (1000.0 * 3600.0);
+
+  for (PetscInt icell = 0; icell < ncells; icell++) {
+    rain[icell] = rain_rate * mm_per_hr_2_m_per_sec;
+    ;
+  }
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+// open the binary file
+static PetscErrorCode OpenSpatiallyHeterogeneousRainData(HeterogeneousRainData *hetero_rain) {
+  PetscFunctionBegin;
+  snprintf(hetero_rain->file, PETSC_MAX_PATH_LEN - 1, "%s/%4d-%02d-%02d:%02d-%02d.%s.bin", hetero_rain->dir, hetero_rain->current_date.yy,
+           hetero_rain->current_date.mo, hetero_rain->current_date.dd, hetero_rain->current_date.hh, hetero_rain->current_date.mm, PETSC_ID_TYPE);
+
+  printf("Opening %s \n", hetero_rain->file);
+  hetero_rain->dtime_in_hour = 1.0;  // assume an hourly dataset
+  hetero_rain->ndata_file    = 1;
+
+  PetscCall(OpenData(hetero_rain->file, &hetero_rain->data_vec, &hetero_rain->ndata));
+  PetscCall(VecGetArray(hetero_rain->data_vec, &hetero_rain->data_ptr));
+
+  hetero_rain->header_offset = 5;
+
+  hetero_rain->ncols    = (PetscInt)hetero_rain->data_ptr[0];
+  hetero_rain->nrows    = (PetscInt)hetero_rain->data_ptr[1];
+  hetero_rain->xlc      = hetero_rain->data_ptr[2];
+  hetero_rain->ylc      = hetero_rain->data_ptr[3];
+  hetero_rain->cellsize = hetero_rain->data_ptr[4];
+
+  if (0) {
+    printf("ncols = %d\n", hetero_rain->ncols);
+    printf("nrows = %d\n", hetero_rain->nrows);
+    printf("xlc   = %f\n", hetero_rain->xlc);
+    printf("ylc   = %f\n", hetero_rain->ylc);
+    printf("size  = %f\n", hetero_rain->cellsize);
+  }
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+// increment the date by one hour
+static PetscErrorCode IncrementDateByAnHour(Date *date) {
```

Comment:
This function isn't leap-year safe. If we use C's `tm` struct, [we can implement this much more easily](https://stackoverflow.com/questions/15668638/arithmetics-on-calendar-dates-in-c-or-c-add-n-days-to-given-date). See "The C Approach" in the accepted answer.

Reply:
Of course, we are probably subject to the constraints of interacting with E3SM...